core/src/main/java/org/apache/calcite/rel/rules/JoinConditionOrExpansionRule.java

```diff
@@ -0,0 +1,297 @@
+    joins.add(expandSemiOrAntiJoin(join, JoinRelType.SEMI, orConds, relBuilder));
+    relBuilder.pushAll(joins)
+        .union(true, joins.size());
+
+    final List<RexNode> projects = join.getRowType().getFieldList().stream()
+        .map(field -> {
+          RexNode rexNode = relBuilder.field(field.getIndex());
+          return field.getType().equals(rexNode.getType())
+              ? rexNode
+              : relBuilder.getRexBuilder().makeCast(field.getType(), rexNode, true, false);
+        }).collect(Collectors.toList());
+
+    return relBuilder.project(projects)
+        .build();
+  }
+
+  /**
+   * This method will make the following conversions.
+   *
+   * <p>Project[*]
+   *    └── Join[OR(id=id0, age=age0), inner]
+   *        ├── TableScan[tbl]
+   *        └── TableScan[tbl]
+   *
+   * <p>into
+   *
+   * <p>Project[*]
+   *    └── UnionAll
+   *        ├── HashJoin[id=id0, inner]
+   *        │   ├── TableScan[tbl]
+   *        │   └── TableScan[tbl]
+   *        └── HashJoin[age=age0 AND id≠id0, inner]
+   *            ├── TableScan[tbl]
+   *            └── TableScan[tbl]
+   */
+  private RelNode expandInnerJoin(Join join, List<RexNode> orConds,
+      RelBuilder relBuilder) {
+    List<RelNode> joins = expandInnerJoinToRelNodes(join, orConds, relBuilder);
+    return relBuilder.pushAll(joins)
+        .union(true, joins.size())
+        .build();
+  }
+
+  private List<RelNode> expandInnerJoinToRelNodes(Join join, List<RexNode> orConds,
+      RelBuilder relBuilder) {
+    List<RelNode> joins = new ArrayList<>();
+    for (int i = 0; i < orConds.size(); i++) {
+      RexNode orCond = orConds.get(i);
+      for (int j = 0; j < i; j++) {
+        orCond = relBuilder.and(orCond, relBuilder.not(orConds.get(j)));
+      }
+      joins.add(relBuilder.push(join.getLeft())
+          .push(join.getRight())
+          .join(JoinRelType.INNER, orCond)
+          .build());
+    }
+    return joins;
+  }
+
+  /**
+   * This method will make the following conversions.
+   *
+   * <p>Project[*]
+   *    └── Join[OR(id=id0, age=age0), anti]
+   *        ├── TableScan[tbl]
+   *        └── TableScan[tbl]
+   *
+   * <p>into
+   *
+   * <p>HashJoin[id=id0, anti]
+   *    ├── HashJoin[age=age0, anti]
+   *    │   ├── TableScan[tbl]
+   *    │   └── TableScan[tbl]
+   *    └── TableScan[tbl]
+   */
+  private RelNode expandSemiOrAntiJoin(Join join, JoinRelType joinType,
+      List<RexNode> orConds, RelBuilder relBuilder) {
+    RelNode left = join.getLeft();
+    for (int i = 0; i < orConds.size(); i++) {
+      RexNode orCond = orConds.get(i);
+      relBuilder.push(left)
+          .push(join.getRight())
+          .join(joinType, orCond);
+      left = relBuilder.build();
+    }
+
+    relBuilder.push(left);
+    int fieldCount = join.getRowType().getFieldCount();
+    List<RexNode> projects = new ArrayList<>(relBuilder.fields());
+    while (fieldCount > relBuilder.fields().size()) {
+      projects.add(relBuilder.literal(null));
```

Comment:
does the builder infer types for these?

Reply:
Builder can not do it. It is a SEMI/ANTI JOIN, there are fewer columns than INNER JOIN, and the columns will not match during UNION. NULL is used here to fill in the gaps, and the JOIN will infer types using derive method later.

Reply:
but can you help the builder by supplying nulls with the correct type?

Reply:
I think I understand what you mean. Using literal directly may cause potential problems in some special cases.

Reply:
If you already know the type, it's probably best to use it rather than infer it again

Reply:
Yes, maybe I can know the types.

Reply:
Done